lazy_tensor_core/dispatch_lazy_mode.py

```diff
@@ -0,0 +1,54 @@
+import torch
+
+import lazy_tensor_core
+lazy_tensor_core._LAZYC._ltc_init_ts_backend()
+import lazy_tensor_core.debug.metrics as metrics
+import lazy_tensor_core.core.lazy_model as ltm
+
+torch._C._jit_set_nvfuser_enabled(True)
+torch._C._jit_set_texpr_fuser_enabled(False)
+torch._C._jit_set_profiling_executor(True)
+torch._C._jit_set_profiling_mode(True)
+torch._C._jit_override_can_fuse_on_cpu(False)
+torch._C._jit_override_can_fuse_on_gpu(False)
+torch._C._jit_set_bailout_depth(20)
+
+class lazy_execute(object):
+    def __init__(self, enabled=True):
+        self._enabled = enabled
+        
+    def __enter__(self):
+        lazy_tensor_core._LAZYC._ltc_enable_lazy_mode()
+               
+    def __exit__(self, type, value, traceback) : 
```

Comment:
What do you want the UX of this to look like at the boundary? We have a couple of options. My main questions are:
1. What happens to `a` and `b` when we enter torch.lazy_execute()? What about when they are used under the context manager?
2. What happens to `c` when we exit the context manager?

```
a = torch.randn(3)
b = torch.randn(3)
with torch.lazy_execute():
  # (A)
  c = a + b

# (B)
c
```

Reply:
Our options for (B) are:
- c might be a "LazyTensor" that the user then has to call .cpu() / .cuda() on for it to be useful
- c might be a regular PyTorch Tensor (that is not Lazy)!

It's... kind of interesting to get the latter to work correctly

Reply:
I think the ideal scenario for (B) is that c is a regular eager tensor after we exit the mode scope.

A possible way to implement this is
 - we assume/assert that the only Lazy tensors are the ones that got created inside _this_ lazy mode, so 'all live lazy tensors' == 'the lazy tensors we might want to use again in eager scope after mode exit'
 - we iterate all the live lazy tensors at exit, and move them to {lazy_eager_device}, which could be cuda or cpu depending on the backend configuration
 - this 'to' operation would have some overhead (constructing new tensor wrappers) but should be copyless (for the underlying data), and may be efficient enough?
 - we'd have to see how many live tensors there typically are.  internal tensors representing intermediate values of layers would likely already be dead, but in some cases there could be a lot of live tensors.

Reply:
At (A) we have a Tensor with an LTCTensorImpl (with a secret Tensor + TensorImpl + cuda storage that is hidden inside)
At (B) we want to have a Tensor with a TensorImpl with cuda storage

you can not swap a tensorimpl underneath a tensor.  you have to modify TensorImpl to add some kind of optional[LtcMetadata]

why can't we just call something like `c = c.to(cuda)` on exit?
- what if c got stashed 'somewhere else' like a global list
- can we write an __exit__ function that rewrites the value of `c`?

We know which lazy tensors are live, and we (maybe) ok with converting them all to eager/cuda on exit
- could we get lists of pyobjects in c++ that represent these live lazy tensors and swap them out?
- the pyobject is a 'THPVariable' which has an at::Tensor associated with it
- if it's a subclass of torch.Tensor then it may have python state and we may get bit there

Reply:
what if at (B) c is just an LTCTensorImpl, 
and we make some special dispatch key that lets us construct an eager tensor?
(this is similar to what is done in functorch, it works but not in all cases)

Reply:
> what if at (B) c is just an LTCTensorImpl,
and we make some special dispatch key that lets us construct an eager tensor?
(this is similar to what is done in functorch, it works but not in all cases)

regarding approach 3.

Tricky case with things that doesn't go through dispatch .e.g. `data_ptr()` this will break as we are accessing TensorImpl::data_ptr()